File: client/src/fabric/twist.ts
import { Vector3 } from "three"

import { avg, midpoint, PHI, pointsToNormal, ROOT3, ROOT6, sub } from "./eig-util"
import { Tensegrity } from "./tensegrity"
import {
    areAdjacent,
    FaceName,
    factorFromPercent,
    IFace,
    IInterval,
    IJoint,
    IPercent,
    IRole,
    Spin,
} from "./tensegrity-types"

const PULL_A: IRole = {
    tag: "(a)",
    push: false,
    length: 1,
    stiffness: 1,
}

const PULL_B: IRole = {
    tag: "(b)",
    push: false,
    length: ROOT3,
    stiffness: 1,
}

const PUSH_B: IRole = {
    tag: "[B]",
    push: true,
    length: PHI * ROOT3,
    stiffness: 1,
}

const PUSH_A: IRole = {
    tag: "[A]",
    push: true,
    length: ROOT6,
    stiffness: 1,
}

export class Twist {

    public readonly faces: IFace[] = []
    public readonly pushes: IInterval[] = []
    public readonly pulls: IInterval[] = []

    constructor(
        public readonly tensegrity: Tensegrity,
        public readonly spin: Spin,
        public readonly scale: IPercent,
        baseKnown?: Vector3[],
    ) {
        const base = !baseKnown ? createBase(new Vector3(), 3) :
            baseKnown.length === 3 ? baseKnown : createBase(baseKnown[0], 3)
        switch (this.spin) {
            case Spin.Left:
                this.createSingle(base, true)
                break
            case Spin.Right:
                this.createSingle(base, false)
                break
            case Spin.LeftRight:
                this.createDouble(base, true)
                break
            case Spin.RightLeft:
                this.createDouble(base, false)
                break
            default:
                throw new Error("Spin?")
        }
    }

    public face(faceName: FaceName): IFace {
        switch (this.faces.length) {
            case 2:
                switch (faceName) {
                    case FaceName.a:
                        return this.faces[0]
                    case FaceName.A:
                        return this.faces[1]
                }
                break
            case 8:
                switch (faceName) {
                    case FaceName.a:
                        return this.faces[0]
                    case FaceName.B:
                        return this.faces[2]
                    case FaceName.C:
                        return this.faces[1]
                    case FaceName.D:
                        return this.faces[3]
                    case FaceName.b:
                        return this.faces[4]
                    case FaceName.c:
                        return this.faces[5]
                    case FaceName.d:
                        return this.faces[6]
                    case FaceName.A:
                        return this.faces[7]
                }
                break
        }
        throw new Error(`Face ${FaceName[faceName]} not found in twist with ${this.faces.length} faces`)
    }

    public get adjacentPulls(): IInterval[] {
        return this.tensegrity.intervals
            .filter(({role}) => !role.push)
            .reduce((detailsSoFar, interval) => {
                const adjacent = this.pushes.find(push => areAdjacent(push, interval))
                if (adjacent) {
                    detailsSoFar.push(interval)
                }
                return detailsSoFar
            }, [] as IInterval[])
    }

    private createSingle(base: Vector3[], leftSpin: boolean): void {
        const pairs = pointPairs(base, this.scale, leftSpin)
        const ends = pairs.map(({alpha, omega}) => ({
            alpha: this.tensegrity.createJoint(alpha),
            omega: this.tensegrity.createJoint(omega),
        }))
        const alphaJoint = this.tensegrity.createJoint(midpoint(pairs.map(({alpha}) => alpha)))
        const omegaJoint = this.tensegrity.createJoint(midpoint(pairs.map(({omega}) => omega)))
        this.tensegrity.instance.refreshFloatView()
        ends.forEach(({alpha, omega}) => {
            const push = this.tensegrity.createInterval(alpha, omega, PUSH_A, this.scale)
            this.pushes.push(push)
            alpha.push = omega.push = push
        })
        const makeFace = (joints: IJoint[], midJoint: IJoint) => {
            const pulls = joints.map(j => this.tensegrity.createInterval(j, midJoint, PULL_A, this.scale))
            this.pulls.push(...pulls)
            this.faces.push(this.tensegrity.createFace(this, joints, pulls, this.spin, this.scale, midJoint))
        }
        makeFace(ends.map(({alpha}) => alpha), alphaJoint)
        makeFace(ends.map(({omega}) => omega).reverse(), omegaJoint)
        ends.forEach(({alpha}, index) => {
            const omega = ends[(ends.length + index + (leftSpin ? -1 : 1)) % ends.length].omega
            this.pulls.push(this.tensegrity.createInterval(alpha, omega, PULL_B, this.scale))
        })
    }

    private createDouble(base: Vector3[], leftSpin: boolean): void {
        const botPairs = pointPairs(base, this.scale, leftSpin)
        const topPairs = pointPairs(botPairs.map(({omega}) => omega), this.scale, !leftSpin)
        const bot = botPairs.map(({alpha, omega}) => ({
            alpha: this.tensegrity.createJoint(alpha),
            omega: this.tensegrity.createJoint(omega),
        }))
        const top = topPairs.map(({alpha, omega}) => ({
            alpha: this.tensegrity.createJoint(alpha),
            omega: this.tensegrity.createJoint(omega),
        }))
        this.tensegrity.instance.refreshFloatView()
        const ends = [...bot, ...top]
        ends.forEach(({alpha, omega}) => {
            const push = this.tensegrity.createInterval(alpha, omega, PUSH_B, this.scale)
            this.pushes.push(push)
            alpha.push = omega.push = push
        })
        const faceJoints = leftSpin ?
            [
                [bot[0].alpha, bot[1].alpha, bot[2].alpha], // a
                [bot[2].alpha, bot[1].omega, top[1].alpha], // B
                [bot[0].alpha, bot[2].omega, top[2].alpha], // C
                [bot[1].alpha, bot[0].omega, top[0].alpha], // D
                [top[1].omega, top[0].alpha, bot[1].omega].reverse(), // b
                [top[0].omega, top[2].alpha, bot[0].omega].reverse(), // d
                [top[2].omega, top[1].alpha, bot[2].omega].reverse(), // c
                [top[0].omega, top[1].omega, top[2].omega].reverse(), // A
            ] : [
                [bot[0].alpha, bot[1].alpha, bot[2].alpha], // a
                [bot[2].alpha, bot[0].omega, top[2].alpha].reverse(), // D
                [bot[0].alpha, bot[1].omega, top[0].alpha].reverse(), // B
                [bot[1].alpha, bot[2].omega, top[1].alpha].reverse(), // C
                [top[1].omega, top[2].alpha, bot[2].omega], // b
                [top[2].omega, top[0].alpha, bot[0].omega], // c
                [top[0].omega, top[1].alpha, bot[1].omega], // d
                [top[0].omega, top[1].omega, top[2].omega].reverse(), // A
            ]
        const jointLocation = (joint: IJoint) => this.tensegrity.instance.jointLocation(joint)
        const midJoints = faceJoints.map(joints => this.tensegrity.createJoint(midpoint(joints.map(jointLocation))))
        this.tensegrity.instance.refreshFloatView()
        faceJoints.forEach((joints, index) => {
            const midJoint = midJoints[index]
            const pulls = joints.map(j => this.tensegrity.createInterval(j, midJoint, PULL_A, this.scale))
            this.pulls.push(...pulls)
            const spin = leftSpin === ([0, 4, 5, 6].some(n => n === index)) ? Spin.Left : Spin.Right
            this.faces.push(this.tensegrity.createFace(this, joints, pulls, spin, this.scale, midJoint))
        })
    }
}

interface IPointPair {
    alpha: Vector3
    omega: Vector3
}

function pointPairs(base: Vector3[], scale: IPercent, leftSpin: boolean): IPointPair[] {
    const points: IPointPair[] = []
    const mid = midpoint(base)
    const midVector = () => new Vector3().copy(mid)
    const factor = factorFromPercent(scale)
    const up = pointsToNormal(base).multiplyScalar(-factor)
    for (let index = 0; index < base.length; index++) {
        const fromMid = (offset: number) => sub(base[(index + base.length + offset) % base.length], mid)
        const between = (idx1: number, idx2: number) => avg(fromMid(idx1), fromMid(idx2))
        const alpha = midVector().addScaledVector(between(0, 1), factor)
        const omega = midVector().add(up).addScaledVector(leftSpin ? between(1, 2) : between(-1, 0), factor)
        points.push({alpha, omega})
    }
    return points
}

function createBase(location: Vector3, pushesPerTwist: number): Vector3[] {
    const base: Vector3[] = []
    for (let index = 0; index < pushesPerTwist; index++) {
        const angle = index * Math.PI * 2 / pushesPerTwist
        const x = Math.cos(angle)
        const y = Math.sin(angle)
        base.push(new Vector3(x, 0, y).add(location))
    }
    return base
}

File: eig/src/twist.rs
use std::f32::consts::PI;

use cgmath::{EuclideanSpace, InnerSpace, Point3, Vector3};

use crate::fabric::{Fabric, UniqueId};
use crate::face::Face;
use crate::interval::Role::{*};
use crate::tenscript::{FaceName, Spin};
use crate::tenscript::FaceName::{*};
use crate::tenscript::Spin::{Left, Right};
use crate::tenscript::TenscriptNode;
use crate::tenscript::TenscriptNode::{Branch, Grow, Mark};

const ROOT3: f32 = 1.732_050_8;
const ROOT5: f32 = 2.236_068;
const ROOT6: f32 = 2.449_489_8;
const PHI: f32 = (1f32 + ROOT5) / 2f32;

impl Fabric {
    pub fn single_twist(&mut self, spin: Spin, pretenst_factor:f32, scale_factor: f32, face_id: Option<UniqueId>) -> [(FaceName, UniqueId); 2] {
        let face = face_id.map(|id| self.face(id));
        let scale = face.map(|Face { scale, .. }| *scale).unwrap_or(1.0) * scale_factor;
        let base = self.base_triangle(face);
        let pairs = create_pairs(base, spin, scale, scale);
        let ends = pairs
            .map(|(alpha, omega)|
                (self.create_joint(alpha), self.create_joint(omega)));
        let push_intervals = ends.map(|(alpha, omega)| {
            self.create_interval(alpha, omega, TwistPush, scale * ROOT6 * pretenst_factor)
        });
        let alpha_joint = self.create_joint(middle(pairs.map(|(alpha, _)| alpha)));
        let omega_joint = self.create_joint(middle(pairs.map(|(_, omega)| omega)));
        let alphas_x = ends.map(|(alpha, _)| alpha);
        let alphas = [alphas_x[2], alphas_x[1], alphas_x[0]];
        let alpha_radials = alphas.map(|alpha| {
            self.create_interval(alpha_joint, alpha, Pull, scale)
        });
        let a_minus_face = self.create_face(scale, spin, alpha_radials, push_intervals);
        let omegas: [usize; 3] = ends.map(|(_, omega)| omega);
        let omega_radials = omegas.map(|omega| {
            self.create_interval(omega_joint, omega, Pull, scale)
        });
        let a_plus_face = self.create_face(scale, spin, omega_radials, push_intervals);
        for index in 0..=2 {
            let offset = match spin { Left => 1, Right => -1 };
            let alpha = ends[index as usize].0;
            let omega = ends[(ends.len() as isize + index + offset) as usize % ends.len()].1;
            self.create_interval(alpha, omega, TwistVerticalPull, ROOT3 * scale);
        }
        if let Some(id) = face_id { self.faces_to_loop(id, a_minus_face) }
        [(Aneg, a_minus_face), (Apos, a_plus_face)]
    }

    pub fn double_twist(&mut self, spin: Spin, pretenst_factor:f32, scale_factor: f32, face_id: Option<UniqueId>) -> [(FaceName, UniqueId); 8] {
        let face = face_id.map(|id| self.face(id));
        let scale = face.map(|Face { scale, .. }| *scale).unwrap_or(1.0) * scale_factor;
        let base = self.base_triangle(face);
        let widening = 1.5f32;
        let bottom_pairs = create_pairs(base, spin, scale, scale * widening);
        let top_pairs = create_pairs(bottom_pairs.map(|(_, omega)| omega), spin.opposite(), widening, scale);
        let bot = bottom_pairs.map(|(alpha, omega)|
            (self.create_joint(alpha), self.create_joint(omega))
        );
        let top = top_pairs.map(|(alpha, omega)|
            (self.create_joint(alpha), self.create_joint(omega))
        );
        let bot_push = bot.map(|(alpha, omega)| {
            self.create_interval(alpha, omega, DoubleTwistPush, PHI * ROOT3 * scale * pretenst_factor)
        });
        let top_push = top.map(|(alpha, omega)| {
            self.create_interval(alpha, omega, DoubleTwistPush, PHI * ROOT3 * scale * pretenst_factor)
        });
        let face_definitions = match spin {
            Left => [
                (Aneg, Left, [bot[2].0, bot[1].0, bot[0].0], [bot_push[0], bot_push[2], bot_push[1]]),
                (Bpos, Right, [bot[0].0, bot[1].1, top[0].0], [bot_push[0], bot_push[1], top_push[0]]),
                (Cpos, Right, [bot[1].0, bot[2].1, top[1].0], [bot_push[1], bot_push[2], top_push[1]]),
                (Dpos, Right, [bot[2].0, bot[0].1, top[2].0], [bot_push[2], bot_push[0], top_push[2]]),
                (Bneg, Left, [top[2].0, top[1].1, bot[2].1], [top_push[2], top_push[1], bot_push[2]]),
                (Cneg, Left, [top[0].0, top[2].1, bot[0].1], [top_push[0], top_push[2], bot_push[0]]),
                (Dneg, Left, [top[1].0, top[0].1, bot[1].1], [top_push[1], top_push[0], bot_push[1]]),
                (Apos, Right, [top[0].1, top[1].1, top[2].1], [top_push[0], top_push[1], top_push[2]]),
            ],
            Right => [
                (Aneg, Right, [bot[2].0, bot[1].0, bot[0].0], [bot_push[0], bot_push[2], bot_push[1]]),
                (Bpos, Left, [bot[0].0, top[2].0, bot[2].1], [bot_push[0], top_push[2], bot_push[2]]),
                (Cpos, Left, [bot[2].0, top[1].0, bot[1].1], [bot_push[2], top_push[1], bot_push[1]]),
                (Dpos, Left, [bot[1].0, top[0].0, bot[0].1], [bot_push[1], top_push[0], bot_push[0]]),
                (Bneg, Right, [top[0].0, bot[1].1, top[1].1], [top_push[0], bot_push[1], top_push[1]]),
                (Cneg, Right, [top[2].0, bot[0].1, top[0].1], [top_push[2], bot_push[0], top_push[0]]),
                (Dneg, Right, [top[1].0, bot[2].1, top[2].1], [top_push[1], bot_push[2], top_push[2]]),
                (Apos, Left, [top[0].1, top[1].1, top[2].1], [top_push[0], top_push[1], top_push[2]]),
            ],
        };
        let faces = face_definitions
            .map(|(name, spin, indexes, push_intervals)| {
                let middle = middle(indexes.map(|index| self.joints[index].location));
                let mid_joint = self.create_joint(middle);
                let radial_intervals = indexes
                    .map(|outer| self.create_interval(mid_joint, outer, RadialPull, scale));
                let face = self.create_face(scale, spin, radial_intervals, push_intervals);
                (name, face)
            });
        if let Some(id) = face_id { self.faces_to_loop(id, faces[0].1) }
        faces
    }

    pub fn faces_to_loop(&mut self, face_a_id: UniqueId, face_b_id: UniqueId) {
        let (face_a, face_b) = (self.face(face_a_id), self.face(face_b_id));
        let scale = (face_a.scale + face_b.scale) / 2.0;
        let (a, b) = (face_a.radial_joints(self), face_b.radial_joints(self));
        for (alpha, omega) in [(0, 0), (2, 0), (1, 2), (0, 2), (2, 1), (1, 1)] {
            self.create_interval(a[alpha], b[omega], TwistRingPull, scale);
        }
        self.remove_face(face_a_id);
        self.remove_face(face_b_id)
    }

    fn base_triangle(&self, face: Option<&Face>) -> [Point3<f32>; 3] {
        if let Some(face) = face {
            face.radial_joint_locations(&self.joints, self)
        } else {
            [0f32, 2f32, 1f32].map(|index| {
                let angle = index * PI * 2.0 / 3.0;
                Point3::from([angle.cos(), 0.0, angle.sin()])
            })
        }
    }
}

fn find_node(nodes: &[TenscriptNode], sought_face_name: &FaceName) -> Option<TenscriptNode> {
    nodes.iter().find(|node|
        match node {
            Grow { face_name, .. } => {
                face_name == sought_face_name
            }
            Mark { face_name, .. } => {
                face_name == sought_face_name
            }
            Branch { .. } => { false }
        }
    ).cloned()
}

fn create_pairs(base: [Point3<f32>; 3], spin: Spin, alpha_scale: f32, omega_scale: f32) -> [(Point3<f32>, Point3<f32>); 3] {
    let radius_factor = 1.4f32;
    let mid = middle(base).to_vec();
    let up = points_to_normal(base) * (alpha_scale + omega_scale) / -2.0;
    [0, 1, 2].map(|index| {
        let from_mid = |offset| base[(index + 3 + offset) as usize % 3].to_vec() - mid;
        let between = |idx1, idx2| (from_mid(idx1) + from_mid(idx2)) * 0.5 * radius_factor;
        let alpha = mid + between(0, 1) * alpha_scale;
        let offset = match spin { Left => 0, Right => 1 };
        let omega = mid + up + from_mid(offset) * omega_scale;
        (Point3::from_vec(alpha), Point3::from_vec(omega))
    })
}

fn middle(points: [Point3<f32>; 3]) -> Point3<f32> {
    (points[0] + points[1].to_vec() + points[2].to_vec()) / 3f32
}

fn points_to_normal(points: [Point3<f32>; 3]) -> Vector3<f32> {
    let v01 = points[1].to_vec() - points[0].to_vec();
    let v12 = points[2].to_vec() - points[1].to_vec();
    v12.cross(v01).normalize()
}
